pdns/dnsname.hh

```diff
@@ -340,14 +385,33 @@ struct SuffixMatchNode
     d_tree.add(labels, true);
   }
 
+  void remove(const DNSName& name)
+  {
+    d_tree.remove(name);
+  }
+
+  void remove(std::vector<std::string> labels)
+  {
+    d_tree.remove(labels);
+  }
+
   bool check(const DNSName& dnsname) const
   {
     return d_tree.lookup(dnsname) != nullptr;
   }
 
   std::string toString() const
   {
-    return d_human;
+    std::string ret;
+    bool first = true;
+    for (const auto &n: d_tree.getNodes()) {
+      if (!first) {
+        ret += ", ";
+      }
+      first = false;
+      ret += n.toString();
+    }
+    return ret;
```

Comment:
I'm a bit worried about this operation going from `O(1)` to `O(n)`, especially in dnsdist where someone could trigger this by retrieving the list of rules from the `API` for example. Perhaps we could cache the result of this operation to a mutable member variable and invalidate it whenever `::add()` or `::remove()` is called?

Reply:
we could store the dnsnames in a set and refresh `d_human` on `::add()` and `::remove()`

Reply:
Although this means that `::add()` and `::remove()` will be a bit more complex.. I wonder what would be called more often, `toString()` or `::add()`/`::remove()`